File: source/ocaml/src/parallel.ml
let swap arr i j =
  let temp = arr.(i) in
  arr.(i) <- arr.(j);
  arr.(j) <- temp

let map cores init run tasks =
  let tasks = Array.of_list tasks in
  let count = Array.length tasks in

  let next = Atomic.make 0 in
  let order = Array.init count (fun i -> i) in

  let results = Array.make count init in

  let shuffle arr =
    for i = count - 1 downto 1 do
      let j = Random.int (i + 1) in
      swap arr i j
    done
  in
  shuffle order;

  let worker () =
    let rec loop () =
      let index = Atomic.fetch_and_add next 1 in
      if index < count then
        ( let index = order.(index) in
          results.(index) <- run tasks.(index);
          loop () )
    in
    loop ()
  in

  let domains =
    Array.init (cores - 1) (fun _ -> Domain.spawn worker)
  in

  worker ();
  Array.iter Domain.join domains;
  Array.to_list results

File: source/nim/src/parallel.nim
import std/[atomics, random, sequtils]

type SharedWork[T, U] = object
  tasks: ptr seq[T]
  results: ptr seq[U]
  order: ptr seq[int]
  next: ptr Atomic[int]
  count: int
  run: proc(x: T): U {.gcsafe.}

proc worker[T, U](
  work: ptr SharedWork[T, U]
) {.thread.} =

  while true:
    let index = work.next[].fetchAdd(1)
    if index >= work.count: break
    let i = work.order[][index]
    work.results[][i] = work.run(work.tasks[][i])

proc map*[T, U](
  cores: int, 
  run: proc(x: T): U {.gcsafe.}, 
  tasks: seq[T]
): seq[U] =

  let count = tasks.len
  
  var next: Atomic[int]
  var order = toSeq(0..<count)
  order.shuffle()
  
  result = newSeq[U](count)
  
  var work = SharedWork[T, U](
    tasks: unsafeAddr tasks,
    results: addr result,
    order: addr order,
    next: addr next,
    count: count,
    run: run
  )
  
  var threads = newSeq[Thread[ptr SharedWork[T, U]]](cores - 1)
  for t in threads.mitems:
    createThread(t, worker[T, U], addr work)
  
  worker(addr work)
  
  for t in threads:
    joinThread(t)
  